pkg/pullsync/pullstorage/pullstorage.go

```diff
@@ -52,70 +55,87 @@ func New(storer storage.Storer) Storer {
 	}
 }
 
+type intervalFlightResult struct {
+	chs     []swarm.Address
+	topmost uint64
+}
+
 // IntervalChunks collects chunk for a requested interval.
 func (s *ps) IntervalChunks(ctx context.Context, bin uint8, from, to uint64, limit int) (chs []swarm.Address, topmost uint64, err error) {
-	// call iterator, iterate either until upper bound or limit reached
-	// return addresses, topmost is the topmost bin ID
-	var (
-		timer  *time.Timer
-		timerC <-chan time.Time
-	)
-	ch, dbClosed, stop := s.SubscribePull(ctx, bin, from, to)
-	defer func(start time.Time) {
-		stop()
-		if timer != nil {
-			timer.Stop()
-		}
-	}(time.Now())
+	resultC := s.intervalFlight.DoChan(subKey(bin, from, to, limit), func() (interface{}, error) {
+		// call iterator, iterate either until upper bound or limit reached
+		// return addresses, topmost is the topmost bin ID
+		var (
+			timer  *time.Timer
+			timerC <-chan time.Time
+		)
+		ch, dbClosed, stop := s.SubscribePull(context.Background(), bin, from, to)
```

Comment:
you cannot use a `context.Background` here since it leaks. As mentioned before, if all waiting goroutines go away, the singleflight must cancel, otherwise it leaks